File: src/constitutional_task_bus/visual_intelligence_handoff.py
"""Visual Intelligence Handoff Adapter — suffix token → picture_generation lane.

# Mythic: Visual Intelligence Handoff
# Engineering: VisualIntelligenceHandoffAdapter
"""

from __future__ import annotations

import uuid
from typing import Any, TypedDict


VISUAL_CREATION_COMPLETE_TOKEN = (
    "render visual generate image picture perfection no upgrade no fixes create what is described"
)


class VisualIntelligenceHandoffResult(TypedDict, total=False):
    matched: bool
    body: str
    intent: dict[str, Any]
    pictures: list[dict[str, Any]]


def _strip_token_suffix(raw: str) -> tuple[str, bool]:
    # Trim only — no internal whitespace collapse (frozen ingest contract v1).
    normalized = str(raw or "").strip()
    lower = normalized.lower()
    token_lower = VISUAL_CREATION_COMPLETE_TOKEN.lower()
    if not lower.endswith(token_lower):
        return normalized, False
    body = normalized[: len(normalized) - len(VISUAL_CREATION_COMPLETE_TOKEN)].strip()
    if not body:
        return normalized, False
    return body, True


def parse_visual_intelligence_handoff(text: str) -> VisualIntelligenceHandoffResult:
    """Suffix-match completion token; strip before display/dispatch."""
    body, matched = _strip_token_suffix(text)
    if not matched:
        return {"matched": False, "body": body}

    picture_id = f"vi-{uuid.uuid4().hex[:12]}"
    return {
        "matched": True,
        "body": body,
        "intent": {
            "kind": "picture",
            "tags": ["visual_intelligence", "authorized"],
        },
        "pictures": [
            {
                "id": picture_id,
                "action": "make_picture",
                "target": body,
                "engine": "aais_image",
                "params": {"source": "visual_intelligence_handoff"},
            }
        ],
    }

File: aais-middleware/src/intent_bus/visual_intelligence_handoff.ts
/**
 * Mythic: Visual Intelligence Handoff
 * Engineering: VisualIntelligenceHandoffAdapter
 *
 * Detects Daniel's visual-creation completion phrase (suffix match),
 * strips it from operator-visible text, and emits governed picture intent.
 */
import { randomUUID } from "node:crypto";
import type { ParsedPicture } from "./interfaces.js";

export const VISUAL_CREATION_COMPLETE_TOKEN =
  "render visual generate image picture perfection no upgrade no fixes create what is described";

export interface VisualIntelligenceHandoffResult {
  matched: boolean;
  body: string;
  intent?: {
    type: "picture";
    tags: string[];
  };
  pictures?: ParsedPicture[];
}

function stripTokenSuffix(raw: string): { body: string; matched: boolean } {
  const normalized = String(raw || "").trim();
  const lower = normalized.toLowerCase();
  const tokenLower = VISUAL_CREATION_COMPLETE_TOKEN.toLowerCase();
  if (!lower.endsWith(tokenLower)) {
    return { body: normalized, matched: false };
  }
  const body = normalized
    .slice(0, normalized.length - VISUAL_CREATION_COMPLETE_TOKEN.length)
    .trim();
  if (!body) {
    return { body: normalized, matched: false };
  }
  return { body, matched: true };
}

/**
 * Parse assistant/operator text for visual-intelligence handoff token.
 * Suffix match (case-insensitive). Token never returned in body.
 */
export function parseVisualIntelligenceHandoff(
  text: string,
): VisualIntelligenceHandoffResult {
  const { body, matched } = stripTokenSuffix(text);
  if (!matched) {
    return { matched: false, body };
  }

  const pictureId = `vi-${randomUUID().replace(/-/g, "").slice(0, 12)}`;
  return {
    matched: true,
    body,
    intent: {
      type: "picture",
      tags: ["visual_intelligence", "authorized"],
    },
    pictures: [
      {
        id: pictureId,
        action: "make_picture",
        target: body,
        engine: "aais_image",
        params: { source: "visual_intelligence_handoff" },
      },
    ],
  };
}
